src/storage-types/src/connections/aws.rs

```diff
@@ -94,100 +155,214 @@ impl RustType<ProtoAwsCredentials> for AwsCredentials {
-                } else {
-                    external_id_prefix.to_string()
-                };
-                role = role.external_id(external_id);
-            }
-            cred_provider = SharedCredentialsProvider::new(role.build(cred_provider));
+    async fn load_sdk_config_from_credentials(
+        &self,
+        credentials: impl ProvideCredentials + 'static,
+    ) -> Result<SdkConfig, anyhow::Error> {
+        let mut loader = aws_config::from_env().credentials_provider(credentials);
+        if let Some(region) = &self.region {
+            loader = loader.region(Region::new(region.clone()));
         }
-
-        let mut loader = aws_config::from_env()
-            .region(region)
-            .credentials_provider(cred_provider);
         if let Some(endpoint) = &self.endpoint {
             loader = loader.endpoint_url(endpoint);
         }
-        loader.load().await
+        Ok(loader.load().await)
     }
 
-    #[allow(clippy::unused_async)]
     pub(crate) async fn validate(
         &self,
-        _id: GlobalId,
-        _connection_context: &ConnectionContext,
+        id: GlobalId,
+        connection_context: &ConnectionContext,
     ) -> Result<(), anyhow::Error> {
-        Err(anyhow!("Validating SSH connections is not supported yet"))
+        // TODO(mouli): Update this to return user friendly error in
+        // case of failure.
+        let aws_config = self.load_sdk_config(connection_context, id).await?;
+        let sts_client = aws_sdk_sts::Client::new(&aws_config);
+        // TODO(mouli): Update this to return user friendly error in
+        // case of failure.
+        let _ = sts_client.get_caller_identity().send().await?;
+
+        if let AwsAuth::AssumeRole(assume_role) = &self.auth {
+            // Per AWS's recommendation, when validating a connection using
+            // `AssumeRole` authentication, we should ensure that the
+            // role rejects `AssumeRole` requests that don't specify an
+            // external ID.
+            let external_id = None;
+            let credentials = assume_role
+                .dangerously_load_credentials_provider(connection_context, id, external_id)
+                .await?;
+            let aws_config = self.load_sdk_config_from_credentials(credentials).await?;
+            let sts_client = aws_sdk_sts::Client::new(&aws_config);
+            if sts_client.get_caller_identity().send().await.is_ok() {
+                // TODO(mouli): Return a structured `ValidateConnectionError``
+                // instead of `anyhow::Error`, so that we can include additional
+                // details and a hint.
+                bail!(
+                    "AWS connection role does not require an external ID; \
+                     this is INSECURE and allows any Materialize customer \
+                     access to your AWS account"
+                );
+            }
+        }
+
+        Ok(())
     }
 
     pub(crate) fn validate_by_default(&self) -> bool {
```

Comment:
I substantially refactored the code in this file to have more comments, produce better error messages (e.g., there were several instances of `"error!"` with no additional context in user-facing error messages), and to reduce duplication.